File: 4_Ugly_Number_II.py
class Solution:
    """
    @param {int} n an integer.
    @return {int} the nth prime number as description.
    """
    def nthUglyNumber(self, n):
        # write your code here
        import heapq
        if n <= 1:
            return n
        
        # initialize: make a heap with values of 2,3,5
        n -= 1
        key = [2,3,5]
        h = []
        for i in range(3):
            heapq.heappush(h,(key[i],i))
        
        # look at each current max, add that value *2, *3, *5
        value = key[0]
        while n > 0:
            value, level = heapq.heappop(h)
            while level < 3:
                new_value = key[level] * value
                heapq.heappush(h,(new_value, level))
                level += 1
            n -= 1    
        return value

File: 4_Ugly_Number_II.java
class Solution {
    /**
     * @param n an integer
     * @return the nth prime number as description.
     */
    public int nthUglyNumber(int n) {
        // Write your code here
        List<Integer> uglys = new ArrayList<Integer>();
        uglys.add(1);
        
        int p2 = 0, p3 = 0, p5 = 0;
        
        for (int i = 1; i < n; i++){
            
            // make sure the currect number is larger than the previously largest number
            int lastNumber = uglys.get(i-1);
            while (uglys.get(p2) *2 <= lastNumber) p2++;
            while (uglys.get(p3) *3 <= lastNumber) p3++;
            while (uglys.get(p5) *5 <= lastNumber) p5++;
            
            // take the smallest of the current numbers being considered
            uglys.add(Math.min(uglys.get(p2) *2, Math.min(uglys.get(p3) *3, uglys.get(p5) *5)));
        }
        
        return uglys.get(n-1);
    }
};
